module/support/optimisation/NSGA2Optimiser/src/nsga2/NSGA2.hpp

```diff
@@ -0,0 +1,122 @@
+
+#include <fstream>
+#include <limits>
+#include <memory>
+#include <numeric>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "Population.hpp"
+#include "Random.hpp"
+
+namespace nsga2 {
+    class NSGA2 {
+    public:
+        NSGA2() : rand_gen(std::make_shared<RandomGenerator<>>()) {}
+
+        bool initialize_first_generation();
+        void complete_generation_and_advance();
+        void initialize_next_generation();
+        bool has_met_optimisation_terminal_condition();
+
+        // clang-format off
+        void set_seed(const int& _seed) { rand_gen->set_seed(_seed); }
+        void set_real_variable_count(const int& real_vars_) { real_vars = real_vars_; }
+        void set_bin_variable_count(const int& bin_vars_) { bin_vars = bin_vars_; }
+        void set_objective_count(const int& objectives_) { objectives = objectives_; }
+        void set_constraint_count(const int& constraints_) { constraints = constraints_; }
+        void set_population_size(const int& pop_size_) { pop_size = pop_size_; }
+        void set_target_generations(const int& _generations) { generations = _generations; }
+        void set_real_crossover_probability(const double& real_cross_prob_) { real_cross_prob = real_cross_prob_; }
+        void set_bin_crossover_probability(const double& bin_cross_prob_) { bin_cross_prob = bin_cross_prob_; }
+        void set_real_mutation_probability(const double& real_mut_prob_) { real_mut_prob = real_mut_prob_; }
+        void set_bin_mutation_probability(const double& bin_mut_prob_) { bin_mut_prob = bin_mut_prob_; }
+        void set_eta_c(const double& eta_c_) { etaC = eta_c_; }
+        void set_eta_m(const double& eta_m_) { eta_m = eta_m_; }
+        void set_eps_c(const double& eps_c_) { eps_c = eps_c_; }
+        void set_bit_count(const std::vector<int>& bin_bits_) { bin_bits = bin_bits_; }
+        void set_real_var_limits(const std::vector<std::pair<double, double>>& real_limits_) { real_limits = real_limits_; }
+        void set_bin_var_limits(const std::vector<std::pair<double, double>>& bin_limits_) { bin_limits = bin_limits_; }
+        void set_initial_real_vars(const std::vector<double>& init_real_vars_) { initial_real_vars = init_real_vars_; }
+        void set_initial_population_real_vars(const std::vector<std::vector<double>>& supplied_population_real_vars_)
+                                                { supplied_population_real_vars = supplied_population_real_vars_;}
+        void set_supplied_pop(const bool& supplied_pop_) {supplied_pop = supplied_pop_;}
+        // clang-format on
+
+        std::shared_ptr<Population> get_current_pop();
+
+        bool crowd_obj       = true;
+        int report_count     = 0;
+        int bin_mut_count    = 0;
+        int real_mut_count   = 0;
+        int bin_cross_count  = 0;
+        int real_cross_count = 0;
+        int bit_length       = 0;
+
+        int current_gen;
+        int pop_size;
+        int generations;
```

Comment:
initialise all the variables in this folder, doc directives aren't super important here since it might be a bit hard to do